File: tribonacciNumber.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
	int findValue(int n, int memo[])
	{
		int result;
		if (memo[n])
			return memo[n];
		if (n == 1 || n == 2)
			result = 1;
		else if (n <= 0)
			result = 0;
		else
			result = findValue(n - 3, memo) + findValue(n - 1, memo) + findValue(n - 2, memo);
		return result;
	}
	int tribonacci(int n)
	{
		// int memo[n + 1] = {0};
		// return findValue(n, memo);
		int *memo = new int[n + 1]{0};
		return findValue(n, memo);
	}
};

int main(int argc, char const *argv[])
{
	Solution S;
	cout << S.tribonacci(35);
	return 0;
}

File: tribonacciNumber.py
class Solution:
    memo = {0: 0, 1: 1, 2: 1}

    def tribonacci(self, n: int) -> int:
        if n in self.memo:
            return self.memo[n]
        else:
            self.memo[n] = self.tribonacci(
                n-1)+self.tribonacci(n-2)+self.tribonacci(n-3)
            return self.memo[n]


b = Solution()
print(b.tribonacci(37))
